README.md
# Player Assistant
Player Assistant is a **Chrome extension** which acts as an auxiliary tool for **video watching**. Because not every video player provides the same keyboard shortcuts, the goal of Player Assistant is to integrate common shortcuts of these players, make every video player can have the same integrated features, and give users a better viewing experience.

# Installation
### Install through Chrome Web Store
[![install-on-chrome-web-store](./media/install-on-chrome-web-store.png)](https://chrome.google.com/webstore/detail/player-assistant/igdagoipdgdcidbkflnildofndcbnfff?hl=zh-TW&authuser=1)

# Usage
### Enable Player Assistant
Click the video to enable Player Assistant.\
*(Note: Click within the video itself & Player Assistant is activated only when the video is selected)*

    
### Keyboard shortcuts
When Player Assistant is enabled and the video is selected, you can use following shortcuts for a specific action. Also note that some players icon wouldn't react to actions taken by Player Assistant, but that doesn't mean actions haven't been taken.
+ Shortcuts for **normal websites**
    | Shortcut                  | Action                                     |
    |---------------------------|--------------------------------------------|
    | <kbd>a</kbd>              | Skip backward(5 seconds in default)        |
    | <kbd>d</kbd>              | Skip forward(5 seconds in default)         |
    | <kbd>q</kbd>              | Decrease playback speed(0.25x in default)  |
    | <kbd>e</kbd>              | Increase playback speed(0.25x in default)  |
    | <kbd>r</kbd>              | Reset playback speed to 1.0x               |
    | <kbd>s</kbd>              | Decrease volume(10% in default)            |
    | <kbd>w</kbd>              | Increase volume(10% in default)            |
    | <kbd>space</kbd>          | Play/Pause                                 |
    | <kbd>f</kbd>              | Toggle fullscreen                          |
    | <kbd>m</kbd>              | Toggle mute                                |
    | <kbd>home</kbd>           | Go to beginning  of video                  |
    | <kbd>end</kbd>            | Go to end of video                         |
    | <kbd>0</kbd>-<kbd>9</kbd> | Go to 0% - 90% of video                    |

+ **Additional** shortcuts for **Netflix**\
*(Note: All of the original Netflix shortcuts are **still available**. You can still use <kbd>s</kbd> to skip intro or recap.)*
    | Shortcut                  | Action                                                  |
    | --------------------------|---------------------------------------------------------|
    | <kbd>n</kbd>              | Play next episode(if available)                         |
    | <kbd>c</kbd>              | Toggle caption between 'Traditional Chinese' and 'English'(if available) |
    
    

### Popup window
+ When Player Assistant is enabled, you can hover over the progress bars **for the current time/speed/volume**.
    <kbd>![bars](./media/github_bars.gif?raw=true)</kbd>
+ When Player Assistant is enabled, you can click the rightmost buttons, enter the value, and press <kbd>enter</kbd> to **edit time/speed/volume offset**. If the edit is valid, you will see the background color of the button change to yellow for a second, that means the offset has been changed successfully and it will immediately affect every website that Player Assistant is currently enabled.
    <kbd>![buttons](./media/github_buttons.gif?raw=true)</kbd>

### Pin the extension
Pin the extension on Chrome to check the current status by the icon.

| Icon                                               | Description                                                             |
|----------------------------------------------------|-------------------------------------------------------------------------|
|![action_on](./media/github_on.png?raw=true)        |Indicates that Player Assistant has been enabled on the current page     |
|![action_off](./media/github_off.png?raw=true)      |Indicates that Player Assistant has NOT been enabled on the current page |

# License
This project is under the [MIT License](./LICENSE).


CHANGELOG.md
## v1.1.4.2
+ Bug fix for Youtube play/pause issue
## v1.1.4.1
+ Netflix caption toggle bug fixes
## v1.1.4
+ Migrated to Manifest V3
## v1.1.3
+ Auto focus bug fixes
+ Embedded Youtube videos bug fixes
+ Netflix draining next episode button bug fixes
## v1.1.2
+ Netflix next episode button bug fixes
## v1.1.1
+ Netflix bug fixes
## v1.1.0
+ cross-origin videos & Youtube is now available for Player Assistant
+ change default volume offset from 5% to 10%
+ increase reliability and decrease the possibility for keyboard shortcut conflicts
+ fix 's' key conflict(volume down & skipping intro) in Netflix
+ change max playback speed of Netflix to 16x
## v1.0.0
+ initial release


pa4cross-origin.js
'use strict';
if (document.getElementsByTagName('video').length > 0) {
    // event handler for 'keydown'
    const keyboardAction = (event) => { // #functions = 8
        if (event.ctrlKey || event.altKey || event.shiftKey || event.metaKey) {
            event.stopImmediatePropagation();
            return;
        }

        switch (event.code) {
            /***** time forward/rewind *****/
            case 'KeyD':
                pa.currentTime += time_offset;
                break;
            case 'KeyA':
                pa.currentTime -= time_offset;
                break;
            /***** playback speed fast/slow/default *****/
            case 'KeyE':
                pa.playbackRate = getValidValue('playbackrate', pa.playbackRate + speed_offset);
                break;
            case 'KeyQ':
                pa.playbackRate = getValidValue('playbackrate', pa.playbackRate - speed_offset);
                break;
            case 'KeyR':
                pa.playbackRate = 1;
                break;
            /***** volume up/down *****/
            case 'KeyW':
                if (pa.muted) pa.muted = false;
                pa.volume = getValidValue('volume', Math.round((pa.volume + volume_offset) * 100) / 100);
                break;
            case 'KeyS':
                if (pa.muted) pa.muted = false;
                pa.volume = getValidValue('volume', Math.round((pa.volume - volume_offset) * 100) / 100);
                event.stopImmediatePropagation();
                break;
            /***** play/pause *****/
            case 'Space':
                status = pa.paused;
                setTimeout(() => {
                    if (status == pa.paused) {
                        if (pa.paused) pa.play();
                        else pa.pause();
                    }
                }, timeout);
                break;
            /***** enter/exit fullscreen *****/
            case 'KeyF':
                status = document.fullscreen;
                setTimeout(() => {
                    if (status == document.fullscreen) {
                        if (document.fullscreen) document.exitFullscreen();
                        else pa.requestFullscreen();
                    }
                }, timeout);
                break;
            /***** mute/unmute *****/
            case 'KeyM':
                pa.muted = !pa.muted;
                break;
            /***** start/end *****/
            case 'Home':
                pa.currentTime = 0;
                break;
            case 'End':
                pa.currentTime = pa.duration;
                break;
            default: break;
        }

        /***** jump *****/
        if (['0', '1', '2', '3', '4', '5', '6', '7', '8', '9'].includes(event.key)) {
            pa.currentTime = pa.duration * 0.1 * parseInt(event.key);
        }
    }

    const getValidValue = (type, new_value) => {
        let max = 0, min = 0;

        if (type === 'playbackrate') min = 0, max = 16;
        else min = 0, max = 1; // 'volume'

        if (min <= new_value && new_value <= max) {
            return new_value;
        }
        else {
            if (new_value > max) return max;
            else return min;
        }
    }

    const videos = document.getElementsByTagName('video');

    let pa = videos[0];
    let win = pa.ownerDocument.defaultView;

    let time_offset = 5;
    let speed_offset = 0.25;
    let volume_offset = 0.05;

    let status = null;
    let timeout = 200;

    win.addEventListener('click', (event) => {
        event.preventDefault();
    }, false);
    win.addEventListener('keydown', keyboardAction, false);

    if (videos.length == 1) console.log('PA has been enabled.');
    else console.log('PA has been enabled, but multiple videos were found.');
}
else console.log('Please change to another JS context.');

source/js/pa4nf.js
'use strict';
// for locationchange event
history.pushState = (f => function pushState() {
    const ret = f.apply(this, arguments);
    window.dispatchEvent(new Event('locationchange'));
    return ret;
})(history.pushState);

history.replaceState = (f => function replaceState() {
    const ret = f.apply(this, arguments);
    window.dispatchEvent(new Event('locationchange'));
    return ret;
})(history.replaceState);

window.addEventListener('popstate', () => {
    window.dispatchEvent(new Event('locationchange'));
}, false);

// event handler for 'locationchange'
const init = () => {
    window.postMessage({ target: 'inject', msg: 'action', value: 'off' }, "*");
    if (!/https:\/\/(www\.)?netflix\.com\/watch\/.*/.test(window.location.href)) return; // not in the watch page

    let interval_id = setInterval(() => {
        video = document.getElementsByTagName('video')[0];
        if (video !== undefined) {
            window.postMessage({ target: 'inject', msg: 'action', value: 'on' }, "*");
            clearInterval(interval_id);
        }
    }, interval);
}

// event handler for 'keydown'
const keyboardAction = (event) => { // #functions = 7
    try {
        if (!/https:\/\/(www\.)?netflix\.com\/watch\/.*/.test(window.location.href)) return;
        else if (getSessionID() == -1) return;
        else if (event.ctrlKey || event.altKey || event.shiftKey || event.metaKey) return;
        else;

        pa = vp.getVideoPlayerBySessionId(vp.getAllPlayerSessionIds()[getSessionID()]);
        video = document.getElementsByTagName('video')[0];

        switch (event.code) {
            /***** time forward/rewind *****/
            case 'KeyD':
                pa.seek(pa.getCurrentTime() + time_offset);
                break;
            case 'KeyA':
                pa.seek(pa.getCurrentTime() - time_offset);
                break;
            /***** playback speed fast/slow/default *****/
            case 'KeyE':
                video.playbackRate = getValidValue('playbackrate', video.playbackRate + speed_offset);
                break;
            case 'KeyQ':
                video.playbackRate = getValidValue('playbackrate', video.playbackRate - speed_offset);
                break;
            case 'KeyR':
                video.playbackRate = 1;
                break;
            /***** volume up/down *****/
            case 'KeyW':
                pa.setVolume(getValidValue('volume', Math.round((pa.getVolume() + volume_offset) * 100) / 100));
                break;
            case 'KeyS':
                if (document.getElementsByClassName('skip-credits').length != 0) break; // skip intro

                pa.setVolume(getValidValue('volume', Math.round((pa.getVolume() - volume_offset) * 100) / 100));
                break;
            /***** next episode *****/
            case 'KeyN':
                try {
                    const next_btn = document.querySelector('[data-uia="control-next"]') || document.querySelector('[data-uia="next-episode-seamless-button"]') || document.querySelector('[data-uia="next-episode-seamless-button-draining"]');
                    next_btn.click();
                }
                catch (exception) {
                    console.log('[Player Assistant]: No next episode or button has not been loaded.');
                }
                break;
            /***** switch text track *****/
            case 'KeyC':
                const lang_set = ['zh-Hant', 'en'];
                const cur_track_bcp47 = pa.getTextTrack().bcp47;
                for (let tmp_track of pa.getTextTrackList()) {
                    if (tmp_track.bcp47 != cur_track_bcp47 && lang_set.includes(tmp_track.bcp47)) {
                        pa.setTextTrack(tmp_track);
                        break;
                    }
                }
                break;
            /***** start/end *****/
            case 'Home':
                pa.seek(0);
                break;
            case 'End':
                pa.seek(pa.getDuration());
                break;
            default: break;
        }

        /***** jump *****/
        if (['0', '1', '2', '3', '4', '5', '6', '7', '8', '9'].includes(event.key)) {
            pa.seek(pa.getDuration() * 0.1 * parseInt(event.key));
        }
    }
    catch (exception) { }
}

const getSessionID = () => {
    for (let index in vp.getAllPlayerSessionIds()) {
        if (vp.getAllPlayerSessionIds()[index].split('-')[0] == 'watch') {
            return index;
        }
    }

    return -1;
}

const getValidValue = (type, new_value) => {
    let max = 0, min = 0;

    if (type == 'playbackrate') min = 0, max = 16;
    else min = 0, max = 1; // 'volume'

    if (min <= new_value && new_value <= max) {
        return new_value;
    }
    else {
        if (new_value > max) return max;
        else return min;
    }
}

let vp = netflix.appContext.state.playerApp.getAPI().videoPlayer;
let pa = null;
let video = null;

let time_offset = 0;
let speed_offset = 0;
let volume_offset = 0;

const interval = 100;

// receive msg('init'/'ask'/'change offset') from inject
window.addEventListener("message", (event) => {
    if (!event.data.target || event.data.target != 'pa4nf') return;

    if (event.data.msg == "init") {
        let storage = event.data.value;
        time_offset = storage.time_offset * 1000;
        speed_offset = storage.speed_offset;
        volume_offset = storage.volume_offset;
    }
    else if (event.data.msg == "ask") {
        pa = vp.getVideoPlayerBySessionId(vp.getAllPlayerSessionIds()[getSessionID()]);
        video = document.getElementsByTagName('video')[0];

        window.postMessage({
            target: 'inject',
            msg: 'ask',
            value: {
                cur_time: pa.getCurrentTime() / 1000,
                cur_speed: video.playbackRate,
                cur_volume: pa.getVolume(),
                duration: pa.getDuration() / 1000,
                max_speed: 16,
                max_volume: 1,
            }
        }, "*");
    }
    else if (event.data.msg == 'change offset') {
        if (event.data.for == 'time_btn') time_offset = event.data.value * 1000;
        else if (event.data.for == 'speed_btn') speed_offset = event.data.value;
        else volume_offset = event.data.value;
    }
    else console.log('[Player Assistant]: PA4NF received an unknown request.');
}, false);

window.postMessage({ target: 'inject', msg: 'init' }, "*"); // init offset
window.addEventListener('keydown', keyboardAction, false);
window.addEventListener('locationchange', init, false);

init();

source/js/popup.js
'use strict';
document.addEventListener("DOMContentLoaded", () => { // init pb & btn
    chrome.tabs.query({active: true, currentWindow: true}, (tabs) => {
        if (!/https:\/\/(www\.)netflix\.com\/.*/.test(tabs[0].url)) { // not netflix
            chrome.tabs.sendMessage(tabs[0].id, {target: 'pa', msg: 'ask'}, (res) => {
                initProgressBar(res);
            });
        }
        else {
            chrome.tabs.sendMessage(tabs[0].id, {target: 'inject', msg: 'ask'});
            chrome.runtime.onMessage.addListener((req) => {
                if (!req.target || req.target != 'popup') return;

                if (req.msg == 'ask') initProgressBar(req.value);
                else console.log('[Player Assistant]: POPUP received an unknown request.');
            });
        }
    });

    initButton();
});

// functions 4 pb
const initProgressBar = (info) => {
    const time_pb = document.getElementById('time_pb');
    const speed_pb = document.getElementById('speed_pb');
    const volume_pb = document.getElementById('volume_pb');
    // set value
    const time_percentage = Math.round((info.cur_time / info.duration) * 100);
    const speed_percentage = Math.round((info.cur_speed / info.max_speed) * 100);
    const volume_percentage = Math.round((info.cur_volume / info.max_volume) * 100);
    time_pb.setAttribute('value', time_percentage);
    speed_pb.setAttribute('value', speed_percentage);
    volume_pb.setAttribute('value', volume_percentage);
    // set data-label
    time_pb.setAttribute('data-label', getValidValue('time', info.cur_time));
    speed_pb.setAttribute('data-label', getValidValue('speed', info.cur_speed));
    volume_pb.setAttribute('data-label', getValidValue('volume', info.cur_volume));
}

const getValidValue = (type, value) => { // for progree bar
    if (type == 'time') {
        value = parseFloat(value);
        return new Date(value * 1000).toISOString().substr(11, 8);
    }
    else if (type == 'speed') {
        return value + 'x';
    }
    else { // type == 'volume'
        value = parseFloat(value);
        value = Math.round(value * 100);
        value = value.toString() + '%';
        return value;
    }
}

// functions 4 button
const initButton = () => {
    const time_btn = document.getElementById('time_btn');
    const speed_btn = document.getElementById('speed_btn');
    const volume_btn = document.getElementById('volume_btn');
    // get back saved/default value
    chrome.storage.local.get({
        time_offset: 5,
        speed_offset: 0.25,
        volume_offset: 0.10
    }, (storage) => {
        time_btn.innerHTML = storage.time_offset;
        speed_btn.innerHTML = storage.speed_offset;
        volume_btn.innerHTML = storage.volume_offset * 100;
    });
    // adding listener
    time_btn.addEventListener('keyup', onEnter);
    speed_btn.addEventListener('keyup', onEnter);
    volume_btn.addEventListener('keyup', onEnter);
}

const onEnter = (event) => {
    if (event.key == 'Enter') {
        let type = event.target.id;
        let value = parseFloat(event.target.innerHTML);

        if (isValidOffset(type, value)) {
            if (type == 'time_btn') chrome.storage.local.set({time_offset: value});
            else if (type == 'speed_btn') chrome.storage.local.set({speed_offset: value});
            else {
                value /= 100;
                chrome.storage.local.set({volume_offset: value});
            }

            chrome.tabs.query({}, (tabs) => {
                for (let i = 0; i < tabs.length; ++i) {
                    chrome.tabs.sendMessage(tabs[i].id, {
                        target: 'content',
                        msg: 'change offset',
                        for: type,
                        value: value
                    });
                }
            });

            event.target.style.background = 'yellow';
            setTimeout(() => {
                event.target.style.background = '';
            }, 500);
        }
    }    
}

const isValidOffset = (type, value) => { // for changing offset
    if (Number.isNaN(value) || value.toString().length > 5) return false;

    if (type == 'time_btn') {
        if (value <= 0 || value >= 99999) return false;
    }
    else if (type == 'speed_btn') {
        if (value <= 0 || value >= 16) return false;
    }
    else {
        if (value <= 1 || value >= 100) return false;
    }

    return true;
}

source/js/inject.js
'use strict';
// create and append script tag
let script = document.createElement('script');
script.type = 'text/javascript';
script.src = chrome.runtime.getURL('js/pa4nf.js');
document.body.appendChild(script);

// forward msg('init'/'ask'/'change offset') from pa4nf to pa4nf/popup/bg
window.addEventListener("message", (event) => {
    if (!event.data.target || (event.data.target != 'inject' && event.data.target != 'content')) return;

    if (event.data.msg == "init") {
        chrome.storage.local.get({ // get back saved/default value
            time_offset: 5,
            speed_offset: 0.25,
            volume_offset: 0.10
        }, (storage) => {
            window.postMessage({
                target: 'pa4nf',
                msg: 'init',
                value: {
                    time_offset: storage.time_offset,
                    speed_offset: storage.speed_offset,
                    volume_offset: storage.volume_offset
                }
            }, "*");
        });            
    }
    else if (event.data.msg == 'ask') {
        chrome.runtime.sendMessage({
            target: 'popup',
            msg: event.data.msg,
            value: event.data.value
        });
    }
    else if (event.data.msg == 'action') {
        chrome.runtime.sendMessage({
            target: 'bg',
            msg: event.data.msg,
            value: event.data.value
        });
    }
    else console.log('[Player Assistant]: INJECT received an unknown request.');
}, false);

// forward msg('ask'/'change offset') from popup to pa4nf
chrome.runtime.onMessage.addListener((req) => {
    if (!req.target || (req.target != 'inject' && req.target != 'content')) return;

    if (req.msg == 'ask' || req.msg == 'change offset') {
        window.postMessage({target: 'pa4nf', msg: req.msg, for: req.for, value: req.value}, "*");
    }
    else console.log('[Player Assistant]: INJECT received an unknown request.');
})

source/js/background.js
'use strict';
// receive msg('action') from pa/inject
chrome.runtime.onMessage.addListener((req, sender) => {
    if (!req.target || req.target != 'bg') return;

    if (req.msg == 'action') {
        if (req.value == 'on') {
            chrome.action.setPopup({popup: 'popup.html', tabId: sender.tab.id});
            chrome.action.setIcon({path: {"16": "../images/action_on.png"}, tabId: sender.tab.id});
        }
        else if (req.value == 'off') {
            chrome.action.setPopup({popup: '', tabId: sender.tab.id});
            chrome.action.setIcon({path: {"16": "../images/action_off.png"}, tabId: sender.tab.id})
        }
        else console.log('[Player Assistant]: BG received an unknown action request.');
    }
    else console.log('[Player Assistant]: BG received an unknown request.');
})

source/js/pa.js
'use strict';
// event handler for 'fullscreenchange'/'webkitfullscreenchange'
const detectFullscreenChange = () => {
    if (fullscreen_timer) {
        clearTimeout(fullscreen_timer);
        fullscreen_timer = null;
    }
}

// event handler for 'volumechange'
const detectVolumeChange = () => {
    if (mute_timer) {
        clearTimeout(mute_timer);
        mute_timer = null;
    }
}

// event handler for 'play/pause'
const detectPlayPuaseChange = () => {
    if (play_pause_timer) {
        clearTimeout(play_pause_timer);
        play_pause_timer = null;
    }
}

// event handler for 'progress'
const unstuck = () => {
    if (pa.buffered.length > 0 && pa.buffered.start(0) > pa.currentTime) {
        const temp = pa.currentTime;
        pa.currentTime = pa.buffered.start(0);
        pa.currentTime = temp;
    }
}

// event handler for 'keydown'
const keyboardAction = (event) => { // #functions = 8
    if (is_last_mouse_pos_not_within_video || isEditing(event.target) ||
        event.ctrlKey || event.altKey || event.shiftKey || event.metaKey) return;
    
    switch (event.code) {
        /***** time forward/rewind *****/
        case 'KeyD':
            pa.currentTime += time_offset;
            event.stopImmediatePropagation();
            break;
        case 'KeyA':
            pa.currentTime -= time_offset;
            event.stopImmediatePropagation();
            break;
        /***** playback speed fast/slow/default *****/
        case 'KeyE':
            pa.playbackRate = getValidValue('playbackrate', pa.playbackRate + speed_offset);
            event.stopImmediatePropagation();
            break;
        case 'KeyQ':
            pa.playbackRate = getValidValue('playbackrate', pa.playbackRate - speed_offset);
            event.stopImmediatePropagation();
            break;
        case 'KeyR':
            pa.playbackRate = 1;
            event.stopImmediatePropagation();
            break;
        /***** volume up/down *****/
        case 'KeyW':
            if (pa.muted) pa.muted = false; // unmute when adjusting volume

            pa.volume = getValidValue('volume', Math.round((pa.volume + volume_offset) * 100) / 100);
            event.stopImmediatePropagation();
            break;
        case 'KeyS':
            if (pa.muted) pa.muted = false; // unmute when adjusting volume

            pa.volume = getValidValue('volume', Math.round((pa.volume - volume_offset) * 100) / 100);
            event.stopImmediatePropagation();
            break;
        /***** enter/exit fullscreen *****/
        case 'KeyF':
            fullscreen_timer = setTimeout(() => {
                if (document.fullscreen) document.exitFullscreen();
                else pa.requestFullscreen();
            }, timeout);
            break;
        /***** mute/unmute *****/
        case 'KeyM':
            mute_timer = setTimeout(() => {
                pa.muted = !pa.muted;
            }, timeout);
            break;
        /***** play/pause *****/
        case 'Space':
            play_pause_timer = setTimeout(() => {
                if (pa.paused) pa.play();
                else pa.pause();
            }, timeout);
            event.preventDefault(); // prevent from scrolling down
            break;
        /***** start/end *****/
        case 'Home':
            pa.currentTime = 0;
            event.preventDefault(); // prevent from scrolling to the top
            event.stopImmediatePropagation();
            break;
        case 'End':
            pa.currentTime = pa.duration;
            event.preventDefault(); // prevent from scrolling to the bottom
            event.stopImmediatePropagation();
            break;
        default: break;
    }

    /***** jump *****/
    if (['0', '1', '2', '3', '4', '5', '6', '7', '8', '9'].includes(event.key)) {
        pa.currentTime = pa.duration * 0.1 * parseInt(event.key);
    }
}

const getValidValue = (type, new_value) => {
    let max = 0, min = 0;

    if (type == 'playbackrate') min = 0, max = 16;
    else min = 0, max = 1; // 'volume'

    if (min <= new_value && new_value <= max) {
        return new_value;
    }
    else {
        if (new_value > max) return max;
        else return min;
    }
}

const isEditing = (target) => {
    if (target.nodeName == 'INPUT' || target.contentEditable == 'true') {
        return true;
    }
    else return false;
}

// event handler for 'click'
const checkClickedElement = (event) => {
    for (let elem of event.target.ownerDocument.elementsFromPoint(event.clientX, event.clientY)) {
        let video = null;

        if (elem.tagName == 'VIDEO') video = elem;
        else if (elem.getElementsByTagName('video').length == 1) video = elem.getElementsByTagName('video')[0];
        else ;

        if (video && isClickWithinVideo({x: event.clientX, y: event.clientY}, video.getBoundingClientRect())) {
            is_last_mouse_pos_not_within_video = false;
            if (pa != video) { // enable/update pa
                disablePA();
                enablePA(video);
            }
            else console.log('[Player Assistant]: Clicked element is within video scope.');
            break;
        }
        else is_last_mouse_pos_not_within_video = true;
    }  
}

const isClickWithinVideo = (click_coord, video_rect) => {
    if ((video_rect.left <= click_coord.x && click_coord.x <= video_rect.right) &&
        (video_rect.top  <= click_coord.y && click_coord.y <= video_rect.bottom)) {
        return true;
    }
    else return false;
}

const disablePA = () => {
    if (pa) {
        window.removeEventListener('keydown', keyboardAction, true);
        document.removeEventListener('fullscreenchange', detectFullscreenChange, true);
        document.removeEventListener('webkitfullscreenchange', detectFullscreenChange, true);
        pa.removeEventListener('volumechange', detectVolumeChange, true);
        pa.removeEventListener('play', detectPlayPuaseChange, true);
        pa.removeEventListener('pause', detectPlayPuaseChange, true);
        pa.removeEventListener('progress', unstuck, true);

        chrome.runtime.sendMessage({ target: 'bg', msg: 'action', value: 'off' });
    }
}

const enablePA = (video) => {
    pa = video;

    window.addEventListener('keydown', keyboardAction, true);
    document.addEventListener('fullscreenchange', detectFullscreenChange, true);
    document.addEventListener('webkitfullscreenchange', detectFullscreenChange, true);
    pa.addEventListener('volumechange', detectVolumeChange, true);
    pa.addEventListener('play', detectPlayPuaseChange, true);
    pa.addEventListener('pause', detectPlayPuaseChange, true);
    pa.addEventListener('progress', unstuck, true);
    
    setTimeout(() => {
        chrome.runtime.sendMessage({target: 'bg', msg: 'action', value: 'on'});
        console.log('[Player Assistant]: PA has been enabled/updated.');
    }, timeout);
}

let pa = null;

let is_last_mouse_pos_not_within_video = true;

let time_offset = 0;
let speed_offset = 0;
let volume_offset = 0;

const timeout = 300;
let fullscreen_timer = null;
let mute_timer = null;
let play_pause_timer = null;

// receive msg('ask'/'change offset') from popup
chrome.runtime.onMessage.addListener((req, sender, sendResponse) => {
    if (!req.target || (req.target != 'pa' && req.target != 'content')) return;

    if (req.msg == 'ask') {
        sendResponse({
            cur_time: pa.currentTime,
            cur_speed: pa.playbackRate,
            cur_volume: pa.volume,
            duration: pa.duration,
            max_speed: 16,
            max_volume: 1,
        });
    }
    else if (req.msg == 'change offset') {
        if (req.for == 'time_btn') time_offset = req.value;
        else if (req.for == 'speed_btn') speed_offset = req.value;
        else volume_offset = req.value;
    }
    else console.log('[Player Assistant]: PA received an unknown request.');
})

chrome.storage.local.get({ // init offset
    time_offset: 5,
    speed_offset: 0.25,
    volume_offset: 0.10
}, (storage) => {
    time_offset = storage.time_offset;
    speed_offset = storage.speed_offset;
    volume_offset = storage.volume_offset;
});

window.addEventListener('load', () => {
    window.addEventListener('click', checkClickedElement, true);
});